flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/FormatDescriptor.java

```diff
@@ -0,0 +1,108 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.api;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigurationUtils;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/** Describes a format and its options for use with {@link TableDescriptor}. */
+@PublicEvolving
+public final class FormatDescriptor {
+
+    private final String format;
+    private final Map<String, String> options;
+
+    private FormatDescriptor(String format, Map<String, String> options) {
+        this.format = format;
+        this.options = Collections.unmodifiableMap(options);
+    }
+
+    /**
+     * Creates a new {@link Builder} describing a format with the given format identifier.
+     *
+     * @param format The factory identifier for the format.
+     */
+    public static Builder forFormat(String format) {
+        return new Builder(format);
+    }
+
+    public String getFormat() {
+        return format;
+    }
+
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        FormatDescriptor that = (FormatDescriptor) obj;
+        return format.equals(that.format) && options.equals(that.options);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(format, options);
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /** Builder for {@link FormatDescriptor}. */
+    public static class Builder {
+        private final String format;
+        private final Map<String, String> options = new HashMap<>();
+
+        private Builder(String format) {
+            this.format = format;
+        }
+
+        /** Sets the given property on the format. */
+        public <T> Builder option(ConfigOption<T> configOption, T value) {
+            options.put(configOption.key(), ConfigurationUtils.convertValue(value, String.class));
+            return this;
+        }
+
+        /** Sets the given property on the format. */
+        public Builder option(String key, String value) {
```

Comment:
We should make clear in which shape the key needs to be. Including an example. Because in docs such as https://ci.apache.org/projects/flink/flink-docs-master/docs/connectors/table/formats/avro-confluent/#format-options, the prefix is part of the table. But here the prefix is not necessary, right?

Reply:
Yes, it isn't necessary here. In theory we could even be smart enough and not prefix if it is already prefixed (or throw an exception).

Reply:
If we can safely throw an exception, I think that would be great.

Reply:
It's not technically 100% safe, but also not any less safe than all the implicit assumptions Flink already makes on config options. I definitely think a helpful error for a fairly likely user error outweighs the very theoretical case where it could break.